### PR TITLE
feat: node connect

### PR DESCRIPTION
**Pull request by okjodom**
connect to Lnd node, and init simulator

### CODE REVIEW

File: sim-cli/src/main.rs [hunk 1]
```diff
@@ -1,7 +1,9 @@
+use std::collections::HashMap;
 use std::path::PathBuf;
+use std::sync::Arc;
 
 use clap::Parser;
-use sim_lib::Config;
+use sim_lib::{lnd::LndNode, Config, LightningNode, Simulation};
```
Comment (carlaKC) [inline sim-cli/src/main.rs:6]: nittynit: `Simulation` doesn't exist in this commit.

File: sim-lib/src/lib.rs [hunk 2]
```diff
@@ -73,3 +114,30 @@ struct PaymentResult {
     settled: bool,
     action: u8,
 }
+
+pub struct Simulation {
+    // The lightning node that is being simulated.
+    nodes: HashMap<String, Arc<dyn LightningNode>>,
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:120]: Comments are for single-node but structs are for multi node, would just update comments.

**Review by carlaKC — APPROVED** [review]
just nits

File: sim-cli/src/main.rs [hunk 3]
```diff
@@ -15,11 +17,16 @@ async fn main() -> anyhow::Result<()> {
...
-    println!("Simulating...");
-    println!("42 and Done!");
+    let mut clients: HashMap<String, Arc<dyn LightningNode>> = HashMap::new();
 
-    Ok(())
+    // TODO: Connect multiple nodes
+    let cfg = nodes.first().unwrap();
+    let lnd = LndNode::new(cfg.address.clone(), cfg.macaroon.clone(), cfg.cert.clone()).await?;
+
+    clients.insert(cfg.id.clone(), Arc::new(lnd));
```
Comment (sr-gi) [inline sim-cli/src/main.rs:28]:
```suggestion
    for node in nodes {
        let lnd = LndNode::new(node.address, node.macaroon, node.cert).await?;
        clients.insert(node.id, Arc::new(lnd));
    }
```

File: sim-lib/src/lib.rs [hunk 4]
```diff
@@ -1,44 +1,87 @@
...
+    )]
+    pub source: PublicKey,
     // The destination of the action.
-    dest: PublicKey,
+    #[serde(
+        serialize_with = "serialize_public_key",
+        deserialize_with = "deserialize_public_key"
+    )]
+    pub dest: PublicKey,
     // The frequency of the action, as in number of times per minute.
```
Comment (sr-gi) [inline sim-lib/src/lib.rs:40]: I introduced this in a previous PR, but it looks like our frequency is not expressed as times per minutes but as delay between actions. e.g. a frequency of 5 would imply a payment every 5 seconds.

File: sim-lib/src/lib.rs [hunk 5]
```diff
@@ -1,44 +1,87 @@
...
+    {
+        PublicKey::from_str(value).map_err(serde::de::Error::custom)
+    }
+}
+
+fn deserialize_public_key<'de, D>(deserializer: D) -> Result<PublicKey, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    deserializer.deserialize_str(PublicKeyVisitor)
```
Comment (sr-gi) [inline sim-lib/src/lib.rs:74]: I feel like this should not be needed, but let's go with it for now and try to simplify it later

**Review by sr-gi — CHANGES_REQUESTED** [review]

File: sim-lib/src/lib.rs [hunk 6]
```diff
@@ -71,3 +114,30 @@ struct PaymentResult {
...
+    // The lightning node that is being simulated.
+    nodes: HashMap<String, Arc<dyn LightningNode>>,
+
+    // The activity that are to be executed on the node.
+    activity: Vec<ActivityDefinition>,
+}
+
+impl Simulation {
+    pub fn new(
+        nodes: HashMap<String, Arc<dyn LightningNode>>,
```
Comment (carlaKC) [inline sim-lib/src/lib.rs:128]: `String` -> `Pubkey` per @sr-gi's suggestion

File: sim-lib/src/lib.rs (hunk 5 above)
Comment (sr-gi) [reply sim-lib/src/lib.rs:76]: Ok, just add  `features = [ "serde" ]` to the bitcoin crate in `Cargo.toml`

File: sim-cli/src/main.rs [hunk 7]
```diff
@@ -15,11 +17,15 @@ async fn main() -> anyhow::Result<()> {
     let cli = Cli::parse();
 
     let config_str = std::fs::read_to_string(cli.config)?;
-    let config: Config = serde_json::from_str(&config_str)?;
+    let Config { nodes, activity } = serde_json::from_str(&config_str)?;
 
-    println!("Config: {:?}", config);
-    println!("Simulating...");
-    println!("42 and Done!");
+    let mut clients: HashMap<String, Arc<dyn LightningNode>> = HashMap::new();
```
Comment (sr-gi) [inline sim-cli/src/main.rs:22]:
```suggestion
    let mut clients: HashMap<PublicKey, Arc<dyn LightningNode>> = HashMap::new();
```

File: sim-cli/src/main.rs [hunk 8]
```diff
@@ -15,11 +17,15 @@ async fn main() -> anyhow::Result<()> {
...
 
-    Ok(())
+    for node in nodes {
+        let lnd = LndNode::new(node.address, node.macaroon, node.cert).await?;
+        clients.insert(node.id, Arc::new(lnd));
```
Comment (sr-gi) [inline sim-cli/src/main.rs:26]:
```suggestion
        clients.insert(PublicKey::from_str(&node.id)?, Arc::new(lnd));
```

File: sim-cli/src/main.rs [hunk 9]
```diff
@@ -1,7 +1,9 @@
+use std::collections::HashMap;
 use std::path::PathBuf;
+use std::sync::Arc;
```
Comment (sr-gi) [inline sim-cli/src/main.rs:3]: Add 
```
use std::str::FromStr;
use bitcoin::secp256k1::PublicKey;
```

Comment (sr-gi) [reply sim-cli/src/main.rs:3]: You'll need to also import `bitcoin` to this crate

File: sim-lib/src/lib.rs [hunk 10]
```diff
@@ -1,45 +1,48 @@
...
 // Phase 1: Parsed User Input
 
-#[derive(Serialize, Deserialize, Debug)]
+#[derive(Serialize, Deserialize, Debug, Clone)]
 pub struct NodeConnection {
-    id: String,
-    host: String,
-    macaroon: String,
-    cert: String,
+    pub id: String,
```
Comment (sr-gi) [inline sim-lib/src/lib.rs:14]:
```suggestion
    pub id: PublicKey,
```

File: sim-lib/src/lib.rs [hunk 11]
```diff
@@ -72,3 +75,30 @@ struct PaymentResult {
     settled: bool,
     action: u8,
 }
+
+pub struct Simulation {
+    // The lightning node that is being simulated.
+    nodes: HashMap<String, Arc<dyn LightningNode>>,
```
Comment (sr-gi) [inline sim-lib/src/lib.rs:81]:
```suggestion
    nodes: HashMap<PublicKey, Arc<dyn LightningNode>>,
```

File: sim-lib/src/lib.rs (hunk 6 above)
Comment (sr-gi) [reply sim-lib/src/lib.rs:89]: Yep, this is missing

**Review by sr-gi — CHANGES_REQUESTED** [review]
Missing a bunch of `String -> PublicKey` for consistency

File: sim-cli/src/main.rs (hunk 9 above)
Comment (okjodom) [reply sim-cli/src/main.rs:3]: ack

**Review by sr-gi — APPROVED** [review]
LGTM

**Pull request merged**